### PR TITLE
Check out before burning

### PR DESCRIPTION
**Pull request by yrong**
Address the comment in https://github.com/paritytech/polkadot-sdk/pull/2522#discussion_r1430503433

Resolves: SNO-800

### CODE REVIEW

File: parachain/pallets/inbound-queue/src/lib.rs [hunk 1]
```diff
@@ -135,6 +137,9 @@ pub mod pallet {
 
 		/// The upper limit here only used to estimate delivery cost
 		type MaxMessageSize: Get<u32>;
+
+		/// Something to execute an XCM message.
+		type XcmExecutor: ExecuteXcm<Self::RuntimeCall> + XcmAssetTransfers;
```
Comment (vgeddes) [inline parachain/pallets/inbound-queue/src/lib.rs:142]: Lets narrow the scope, and just rely on the transactor:

```suggestion
		type AssetTransactor: TransactAsset;
```

File: parachain/pallets/inbound-queue/src/lib.rs [hunk 2]
```diff
@@ -330,6 +336,33 @@ pub mod pallet {
...
+			let fees = (MultiLocation::parent(), fee.saturated_into::<u128>()).into();
+
+			<T::XcmExecutor as XcmAssetTransfers>::AssetTransactor::can_check_out(
+				&dest,
+				&fees,
+				&dummy_context,
+			)
+			.map_err(|_| Error::<T>::BurnFees)?;
+
+			<T::XcmExecutor as XcmAssetTransfers>::AssetTransactor::check_out(
```
Comment (vgeddes) [inline parachain/pallets/inbound-queue/src/lib.rs:356]:
```suggestion
			T::AssetTransactor::check_out(
```

File: parachain/pallets/inbound-queue/src/lib.rs [hunk 3]
```diff
@@ -330,6 +336,33 @@ pub mod pallet {
...
+		/// Burn the amount of the fee embedded into the XCM for teleports
+		pub fn burn_fees(para_id: ParaId, fee: BalanceOf<T>) -> DispatchResult {
+			let dummy_context =
+				XcmContext { origin: None, message_id: Default::default(), topic: None };
+
+			let dest = MultiLocation { parents: 1, interior: X1(Parachain(para_id.into())) };
+
+			let fees = (MultiLocation::parent(), fee.saturated_into::<u128>()).into();
+
+			<T::XcmExecutor as XcmAssetTransfers>::AssetTransactor::can_check_out(
```
Comment (vgeddes) [inline parachain/pallets/inbound-queue/src/lib.rs:349]:
```suggestion
			T::AssetTransactor::can_check_out(
```

File: parachain/pallets/inbound-queue/src/lib.rs [hunk 4]
```diff
@@ -330,6 +336,33 @@ pub mod pallet {
...
+			.map_err(|_| Error::<T>::BurnFees)?;
+
+			<T::XcmExecutor as XcmAssetTransfers>::AssetTransactor::check_out(
+				&dest,
+				&fees,
+				&dummy_context,
+			);
+
+			let sovereign_account = sibling_sovereign_account::<T>(para_id);
+			T::Token::burn_from(&sovereign_account, fee, Precision::Exact, Fortitude::Polite)?;
```
Comment (vgeddes) [inline parachain/pallets/inbound-queue/src/lib.rs:363]: For consistency, I think we should use `T::AssetTransactor::withdraw_asset`. Thoughts?

Comment (yrong) [reply parachain/pallets/inbound-queue/src/lib.rs]: Nice!

**Comment by yrong** [comment]
https://github.com/Snowfork/snowbridge/pull/1071/commits/7c100640d1a9ed96afc723baba03fdca98882c23 to address the comment https://github.com/paritytech/polkadot-sdk/pull/2522#discussion_r1431464641

File: parachain/pallets/inbound-queue/src/lib.rs [hunk 5]
```diff
@@ -330,6 +338,30 @@ pub mod pallet {
 				.saturating_add(len_fee)
 				.saturating_add(T::PricingParameters::get().rewards.local)
 		}
+
+		/// Burn the amount of the fee embedded into the XCM for teleports
+		pub fn burn_fees(para_id: ParaId, fee: BalanceOf<T>) -> DispatchResult {
+			let dummy_context =
+				XcmContext { origin: None, message_id: Default::default(), topic: None };
```
Comment (alistair-singh) [inline parachain/pallets/inbound-queue/src/lib.rs:345]: Should we add the actual message id and origin here since we have access to this?

**Review by alistair-singh — APPROVED** [review]

File: parachain/pallets/inbound-queue/src/lib.rs [hunk 6]
```diff
@@ -330,6 +338,30 @@ pub mod pallet {
...
+			let dest = MultiLocation { parents: 1, interior: X1(Parachain(para_id.into())) };
+			let fees = (MultiLocation::parent(), fee.saturated_into::<u128>()).into();
```
Comment (alistair-singh) [inline parachain/pallets/inbound-queue/src/lib.rs:347]: Can this be inferred from the `T::Token`?

File: parachain/pallets/inbound-queue/src/mock.rs [hunk 7]
```diff
@@ -221,6 +266,7 @@ impl inbound_queue::Config for Test {
 	type WeightToFee = IdentityFee<u128>;
 	type LengthToFee = IdentityFee<u128>;
 	type MaxMessageSize = ConstU32<1024>;
+	type AssetTransactor = SuccessfulTransactor;
```
Comment (claravanstaden) [inline parachain/pallets/inbound-queue/src/mock.rs:269]: Would be nice to have some tests that checks the `BurnFees` code paths but not a blocker.

**Review by claravanstaden — APPROVED** [review]

File: parachain/pallets/inbound-queue/src/lib.rs (hunk 5 above)
Comment (yrong) [reply parachain/pallets/inbound-queue/src/lib.rs:345]: `XcmContext` is irrelevant in teleports checks so the dummy should be fine.

File: parachain/pallets/inbound-queue/src/mock.rs (hunk 7 above)
Comment (yrong) [reply parachain/pallets/inbound-queue/src/mock.rs:269]: [Change](https://github.com/Snowfork/polkadot-sdk/compare/89a79f518380c55a9dadf3098a025daadb37c08c...452ae5f8fb600b82d4267d519fa51708420bfdf2#diff-5ded4b43c234d6f2d43aff2b2b63adfde7a5ba2dd9fc591cd818335e20b6303b) to cover that.

Also, token-transfer-related smoke tests work as expected.

File: parachain/pallets/inbound-queue/src/lib.rs (hunk 6 above)
Comment (vgeddes) [reply parachain/pallets/inbound-queue/src/lib.rs:347]: Actually, this PR makes it not necessary to have a `T::Token` config field. Can we remove it @ron?

File: parachain/pallets/inbound-queue/src/lib.rs [hunk 8]
```diff
@@ -180,6 +186,8 @@ pub mod pallet {
 		Send(SendError),
 		/// Message conversion error
 		ConvertMessage(ConvertMessageError),
+		/// Burn fees failure
+		BurnFees,
```
Comment (vgeddes) [inline parachain/pallets/inbound-queue/src/lib.rs:190]:
```suggestion
```

Actually we should just raise [TokenError::FundsUnavailable](https://paritytech.github.io/polkadot-sdk/master/sp_runtime/enum.TokenError.html), which I believe is convertible to `DispatchError`.

File: parachain/pallets/inbound-queue/src/lib.rs (hunk 6 above)
Comment (yrong) [reply parachain/pallets/inbound-queue/src/lib.rs:347]: Still in use to reward relayer
https://github.com/Snowfork/snowbridge/blob/f7bd88475c9f52ce19361018de9514b7d0d0979a/parachain/pallets/inbound-queue/src/lib.rs#L270 so I prefer to keep it.

File: parachain/pallets/inbound-queue/src/lib.rs (hunk 8 above)
Comment (yrong) [reply parachain/pallets/inbound-queue/src/lib.rs]: https://github.com/Snowfork/snowbridge/pull/1071/commits/85f5b42b97aa07e825b221a82c9079808a3909ef

**Comment by vgeddes** [comment]
Not a priority to merge this now, can wait until Jan

**Pull request merged**